# Tasks for week 3 in Mathematics 1 TMA4101

## Task 2
Do Dotties experiment in Python: Solve $x=cosx$

In [25]:
#imports used for all tasks
import math as m

In [26]:
"""
Solve x=cos(x) with for-loop
"""
x = 1

for i in range(10):
    x = m.cos(x)
    print(x)

0.5403023058681398
0.8575532158463933
0.6542897904977792
0.7934803587425655
0.7013687736227566
0.7639596829006542
0.7221024250267077
0.7504177617637605
0.7314040424225098
0.7442373549005569


## Task 3
Redo the Dottie experiment but count how many iteration is required to stabilize on $L = 0.739085133215161$

In [27]:
"""
Count iterations for solving x=cos(x) with for-loop
"""
L = 0.7390851332151607
precision = 10
x = 1

for i in range(100):
    if round(x, precision) == round(L, precision):
        print(f"Iterations required: {i}")
        break
    x = m.cos(x)

Iterations required: 57


## Task 4: Solve the Equation Using the Dottie Method

### Objective

The equation $x \ln(x) = 1$ can be rewritten in the form $x = g(x)$ in at least two different ways. Your task is to find these representations and use the Dottie method to find the solutions. 

### Criteria for Success

1. **Find Representations**: Determine the two ways the equation $x \ln(x) = 1$ can be written as $x = g(x)$.
2. **Use Dottie Method**: Employ the Dottie method to find the roots of the equation.
3. **Convergence Criteria**: Continue iterating until the maximal precision $t = 10^{-16}$ is reached.

### Questions

1. What are the two ways you can rewrite the equation $x \ln(x) = 1$ as $x = g(x)$?
 - $x \ln(x) = 1$ can be written as $x = 1 / \ln(x)$ or $x = e^{1/x}$.

2. How many iterations are needed for each method to reach the maximal precision $t = 10^{-16}$?


In [28]:
max_precision = 1e-16
x = 1
prev_x = None

for i in range(1000):
    if x != 0:
        prev_x = x
        x = m.exp(1/x)
        if abs(x - prev_x) < max_precision:
            print(f"Iterations required: {i}")
            print(f"x = {x}")
            break
    else:
        print("x = 0, division by zero")
        break

Iterations required: 63
x = 1.7632228343518968


## Task 5: Solve the Equation Using Newton's Method

### Objective

Solve the equation $x = \cos(x)$ using Newton's Method and count the number of iterations as in Task 3. You may need to rewrite the equation slightly to apply the method effectively.

### Initial Guess ($x_0$)

The first element in the sequence, $x_0$, is known as the initial guess. All numerical solvers require such an initial guess to get started. Some numerical methods are more sensitive to the initial guess than others, and different numerical solvers behave quite differently.
 - Lets rewrite the equation as $x = x - (x - cosx)/(1 + sinx)$ and set $x_0 = 1$.


In [42]:
x = 1 # x_0 = 1
L = 0.7390851332151607
max_precision = 1e-16

for i in range(1000):
    if abs(x - L) < max_precision:
        print(f"Iterations required: {i}")
        print(f"x = {x}")
        break
    
    x = x - (x - m.cos(x))/(1 + m.sin(x))

Iterations required: 5
x = 0.7390851332151607


## Task 6: Solve the Equation $x = \cos(x)$ Using Both Newton's Method and Fixed-Point Iteration

### Objective

Solve the equation $x = \cos(x)$ using both Newton's Method and Fixed-Point Iteration. Experiment with different initial guesses. 

### Questions

1. What happens when $x_0 = 4$ or larger in each of the methods?


In [50]:
L = 0.7390851332151607
max_precision = 1e-16

for x_0 in range(10):
    x = x_0
    y = x_0
    print(f"Initial value: {x_0}")
    for j in range(1000):
        if abs(x - L) < max_precision:
            print(f"    Iterations required for Dottie: {j}")
            break
        
        x = m.cos(x)
    
    for j in range(1000):
        if abs(y - L) < max_precision:
            print(f"    Iterations required for Newton: {j}")
            break
        
        y = y - (y - m.cos(y))/(1 + m.sin(y))

Initial value: 0
    Iterations required for Dottie: 91
    Iterations required for Newton: 5
Initial value: 1
    Iterations required for Dottie: 90
    Iterations required for Newton: 4
Initial value: 2
    Iterations required for Dottie: 92
    Iterations required for Newton: 4
Initial value: 3
    Iterations required for Dottie: 91
    Iterations required for Newton: 7
Initial value: 4
    Iterations required for Dottie: 88
    Iterations required for Newton: 44
Initial value: 5
    Iterations required for Dottie: 92
    Iterations required for Newton: 23
Initial value: 6
    Iterations required for Dottie: 91
    Iterations required for Newton: 10
Initial value: 7
    Iterations required for Dottie: 85
    Iterations required for Newton: 27
Initial value: 8
    Iterations required for Dottie: 92
    Iterations required for Newton: 37
Initial value: 9
    Iterations required for Dottie: 91
    Iterations required for Newton: 6


## Task 7: Repeat for $x = \ln(x)$

### Objective

Repeat the previous tasks for the equation $x = \ln(x)$. Consider both Dottie's Method and Newton's Method to solve the equation and count the number of iterations required for each method with different initial guesses.


In [54]:
max_precision = 1e-16
prev_x = None
prev_y = None

for x_0 in range(10):
    x = x_0
    y = x_0
    print(f"Initial value: {x_0}")
    for j in range(1, 1000):
        prev_x = x
        x = m.cos(x)
        
        if abs(x - prev_x) < max_precision:
            print(f"    Iterations required for Dottie: {j}")
            break
    
    for j in range(1, 1000):
        prev_y = y
        y = y - (y - m.cos(y))/(1 + m.sin(y))
        
        if abs(y - prev_y) < max_precision:
            print(f"    Iterations required for Newton: {j}")
            break

Initial value: 0
    Iterations required for Dottie: 92
    Iterations required for Newton: 6
Initial value: 1
    Iterations required for Dottie: 91
    Iterations required for Newton: 5
Initial value: 2
    Iterations required for Dottie: 93
    Iterations required for Newton: 5
Initial value: 3
    Iterations required for Dottie: 92
    Iterations required for Newton: 8
Initial value: 4
    Iterations required for Dottie: 89
    Iterations required for Newton: 45
Initial value: 5
    Iterations required for Dottie: 93
    Iterations required for Newton: 24
Initial value: 6
    Iterations required for Dottie: 92
    Iterations required for Newton: 11
Initial value: 7
    Iterations required for Dottie: 86
    Iterations required for Newton: 28
Initial value: 8
    Iterations required for Dottie: 93
    Iterations required for Newton: 38
Initial value: 9
    Iterations required for Dottie: 92
    Iterations required for Newton: 7


## Task 8: Find the Water Level in a Sewer Pipe

### Objective

Let $h$ be the water level in a sewer pipe with radius 1 m, and let $h = \cos(\theta)$. When the water fills 75% of the pipe's cross-section, the angle $\theta$ solves the equation $2\sin^2(\theta) + \pi - 4\theta = 0$.

### Questions

1. Find the water level $h$ using Fixed-Point Iteration and Newton's Method.
 - Lets rewrite the equation as $\theta = \theta - \frac{2sin2\theta+\pi-4\theta}{4cos2\theta-4}$ for Newton
 - Lets rewrite the equation as $\theta = \frac{2\sin2(\theta)+\pi}{4}$ for Fixed-Point Iteration

In [68]:
x, y = 1, 1
max_precision = 1e-10

for i in range(1, 100):
    prev_x = x
    x = (2 * m.sin(2 * x) + (m.pi))/(4)
    
    if abs(x - prev_x) < max_precision:
        print(
            f"""
            Dottie:
                Iterations required: {i}
                Thetas: {x}
                h = {round(m.cos(x), 5)}m
            """
        )
        break
    
for i in range(1, 100):
    prev_y = y
    y = y - (2 * m.sin(2 * y) + (m.pi) - 4*y)/(4 * m.cos(2 * y) - 4)
    
    if abs(y - prev_y) < max_precision:
        print(
            f"""
            Newton:
                Iterations required: {i}
                Thetas: {y}
                h = {round(m.cos(y), 5)}m
            """
        )
        break
    



            Dottie:
                Iterations required: 56
                Thetas: 1.1549407299729477
                h = 0.40397m
            

            Newton:
                Iterations required: 5
                Thetas: 1.1549407300050285
                h = 0.40397m
            


## Task 9: Find the Roots of the Equation

### Objective

The equation $x^3 + x^2 + x + 1 = 0$ has one real root and two complex roots. You can find these roots numerically, but you must inform Python that you intend to work with complex numbers by writing the initial guess as $1 + 0 * 1j$.

### Questions

1. Try to find all the roots using numerical methods.


In [111]:
import cmath as cm

# Define functions and initial values
def f(x):
    return x**3 + x**2 + x + 1

def df(x):
    return 3*x**2 + 2*x + 1

# Initial guesses
initial_guesses = [-1, 1, 0 + 1j, 0 - 1j]

precision = 1e-16 # Precision of 16 decimal places (max for float64)


In [112]:

for init_guess in initial_guesses:
    print(f"Initial guess: {init_guess}")
    x = init_guess
    y = init_guess
    # Iterate using Newton's method
    for i in range(1000):
        prev_x = x
        x = x - f(x) / df(x)
        if cm.isinf(x) or cm.isnan(x):
            print("Diverged")
            break
        if abs(x - prev_x) < precision:
            print(f"Root using Newton's Method: {x}, found in {i+1} iterations")
            break
        
    # Iterate using Fixed-point method
    for i in range(1000):
        prev_y = y
        y = x - f(x)
        if cm.isinf(y) or cm.isnan(y):
            print("Diverged")
            break
        if abs(y - prev_y) < precision:
            print(f"Root using Fixed-point Method: {y}, found in {i+1} iterations")
            break

Initial guess: -1
Root using Newton's Method: -1.0, found in 1 iterations
Root using Fixed-point Method: -1.0, found in 1 iterations
Initial guess: 1
Root using Newton's Method: -1.0, found in 10 iterations
Root using Fixed-point Method: -1.0, found in 2 iterations
Initial guess: 1j
Root using Newton's Method: 1j, found in 1 iterations
Root using Fixed-point Method: 1j, found in 1 iterations
Initial guess: -1j
Root using Newton's Method: -1j, found in 1 iterations
Root using Fixed-point Method: -1j, found in 1 iterations


We find the roots $-1$, $j$ and $-j$.

## Task 10: Factorize the Polynomial $x^3 + x^2 - 3x - 3 = 0$

In [142]:
# Define functions and initial values
def f(x):
    return x**3 + x**2 - 3*x - 3

def df(x):
    return 3*x**2 + 2*x - 3

# Round the real and imaginary parts of a complex number to a given precision
def round_complex(z, prec):
    return complex(round(z.real, prec), round(z.imag, prec))

# Initial guesses
initial_guesses = [-1, 1, 1.5, -1.5, 1 + 1j, 1 - 1j, -1 + 1j, -1 - 1j]

# Create a set to keep track of unique roots
unique_roots_newton = set()
unique_roots_fixed = set()

precision = 1e-16 # Precision of 16 decimal places (max for float64)

In [143]:
for init_guess in initial_guesses:
    x = init_guess
    y = init_guess
    
    # Iterate using Newton's method
    for i in range(100):
        prev_x = x
        x = x - f(x) / df(x)
        if cm.isinf(x) or cm.isnan(x):
            break
        if abs(x - prev_x) < precision:
            unique_roots_newton.add(round_complex(x, 10))
            break
    
    # Iterate using Fixed-point method
    for i in range(100):
        prev_y = y
        y = x - f(x)
        if cm.isinf(y) or cm.isnan(y):
            break
        if abs(y - prev_y) < precision:
            unique_roots_fixed.add(round_complex(y, 10))
            break

# Print the unique roots
print(f"Unique roots using Newton's Method: {unique_roots_newton}")
print(f"Unique roots using Fixed-point Method: {unique_roots_fixed}")

Unique roots using Newton's Method: {(-1+0j)}
Unique roots using Fixed-point Method: {(1.7320508076+0j), (-1+0j), (-1.7320508076+0j)}


We find the roots $-1$, $1.73205$ and $-1.73205$.

## Task 11: Find the Roots of Polynomials $ p $ and $ q $

### Objective

The polynomials $ p(x) = x^4 - 10x^3 + 35x^2 - 50x + 24 $ and $ q(x) = x^4 - 9x^3 + 27x^2 - 31x + 12 $ both have a root at $ x = 1 $. Your task is to use Newton's method to find these roots.

### Initial Condition

Start the iteration with an initial guess of $ x_0 = \frac{1}{2} $.

### Questions

1. Note the number of iterations required to find the roots of $ p(x) $ and $ q(x) $ using Newton's method.


In [1]:
# Define functions
def p(x):
    return x**3 - 9*x**2 + 26*x - 24

def dp(x):
    return 3*x**2 - 18*x + 26

def q(x):
    return x**3 - 8*x**2 + 19*x - 12

def dq(x):
    return 3*x**2 - 16*x + 19

# Initial value
x = 0.5

Now, lets use Newtons method to find the roots of $p(x)$ and $q(x)$.

In [6]:
# Initial value
x_p = 0.5  # Initial value for p(x)
x_q = 0.5  # Initial value for q(x)
precision = 1e-10  # Define your stopping criterion

# Find roots for reduced_p(x)
for i in range(1000):
    prev_x = x_p
    x_p = x_p - p(x_p) / dp(x_p)
    if abs(x_p - prev_x) < precision:
        print(f"Root for reduced p(x): {round(x_p, 5)}, found in {i+1} iterations")
        break

# Find roots for reduced_q(x)
for i in range(1000):
    prev_x = x_q
    x_q = x_q - q(x_q) / dq(x_q)
    if abs(x_q - prev_x) < precision:
        print(f"Root for reduced q(x): {round(x_q, 5)}, found in {i+1} iterations")
        break

Root for reduced p(x): 2.0, found in 8 iterations
Root for reduced q(x): 1.0, found in 6 iterations
